### PR TITLE
[SHLWAPI] Initial support of SHAutoComplete

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implement `shlwapi!SHAutoComplete` function.

JIRA issue: [CORE-9281](https://jira.reactos.org/browse/CORE-9281)

## Proposed changes

- Add `autocomp.cpp` and modify `CMakeLists.txt`.
- Implement `IEnumString` body for `SHAutoComplete` and use it in `SHAutoComplete`.
- Rename `shlwapi_main.c` as `shlwapi_main.cpp` and modify it to use ATL.

after

### CODE REVIEW

**Comment by katahiromz** [comment]
BEFORE:
before2
AFTER:
after2
There is no change in `IAutoComplete` body.

**Comment by learn-more** [comment]
Do you intend to send this to wine?
Because if you do not, I can not understand why you would write this in c...

**Comment by katahiromz** [comment]
I will show a friendship with WineHQ about my source. This source can be easily port into WineHQ.

File: dll/win32/shlwapi/autocomp.c [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+
+static AutoComplete_EnumString *
+AutoComplete_EnumString_Construct(SIZE_T capacity)
+{
+    AutoComplete_EnumStringVtbl *lpVtbl;
+    AutoComplete_EnumString *ret = CoTaskMemAlloc(sizeof(AutoComplete_EnumString));
+    if (!ret)
+        return ret;
+
+    lpVtbl = CoTaskMemAlloc(sizeof(AutoComplete_EnumStringVtbl));
```
Comment (learn-more) [inline dll/win32/shlwapi/autocomp.c:494]: The vtable can just be static, compile-time initialized?

Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.c]: OK.

File: dll/win32/shlwapi/autocomp.c [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+#include "assert.h"
+#include "wine/debug.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+#define INITIAL_CAPACITY    256
+#define NUM_GROW            64
+#define MAX_TYPED_URLS      50
+
+typedef struct AutoComplete_EnumStringVtbl
```
Comment (ThFabba) [inline dll/win32/shlwapi/autocomp.c:26]: This struct seems exactly identical to `IEnumStringVtbl`, so there doesn't seem to be a need to redefine it.

File: dll/win32/shlwapi/autocomp.c [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+    ULONG               m_istr;
+    SIZE_T              m_cstrs;
+    SIZE_T              m_capacity;
+    BSTR *              m_pstrs;
+    HWND                m_hwndEdit;
+    DWORD               m_dwSHACF;
+} AutoComplete_EnumString;
+
+static void
+AutoComplete_EnumString_ResetContent(AutoComplete_EnumString *this_)
```
Comment (ThFabba) [inline dll/win32/shlwapi/autocomp.c:60]: Wine's convention is generally to use `This`.

File: dll/win32/shlwapi/autocomp.c [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+    if (!bstr)
+    {
+        ERR("Out of memory\n");
+        return FALSE;
+    }
+
+    if (this_->m_cstrs + 1 >= this_->m_capacity)
+    {
+        new_capacity = this_->m_capacity + NUM_GROW;
+        pstrs = (BSTR *)CoTaskMemAlloc(new_capacity * sizeof(BSTR));
```
Comment (ThFabba) [inline dll/win32/shlwapi/autocomp.c:153]: No need for this cast in C.

File: dll/win32/shlwapi/autocomp.c [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+
+static inline void
+AutoComplete_DoDir(AutoComplete_EnumString *pES, LPCWSTR pszDir, BOOL bDirOnly)
+{
+    LPWSTR pch;
+    WCHAR szPath[MAX_PATH];
+    HANDLE hFind;
+    WIN32_FIND_DATAW find;
+
+    StringCbCopyW(szPath, sizeof(szPath), pszDir);
```
Comment (ThFabba) [inline dll/win32/shlwapi/autocomp.c:284]: Wine doesn't have strsafe.

**Review by ThFabba — COMMENTED** [review]
Not sure if you're aiming for this to be Wine-compatible. If not you can ignore most of my comments (but should also probably use C++ then)

**Comment by katahiromz** [comment]
Colin's answer is to use C++ in our manner. So I will rewrite in C++.

**Comment by katahiromz** [comment]
@learn-more Help me!
```txt
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(188): error C2259: 'CAutoCompleteEnumString': cannot instantiate abstract class
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(26): note: see declaration of 'CAutoCompleteEnumString'
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(188): note: due to following members:
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(188): note: 'HRESULT CAutoCompleteEnumString::QueryInterface(const IID &,void **)': is abstract
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(50): note: see declaration of 'CAutoCompleteEnumString::QueryInterface'
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(188): note: 'ULONG CAutoCompleteEnumString::AddRef(void)': is abstract
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(50): note: see declaration of 'CAutoCompleteEnumString::AddRef'
D:\a\reactos\reactos\src\dll\win32\shlwapi\autocomp.cpp(188): note: 'ULONG CAutoCompleteEnumString::Release(void)': is abstract
```

File: dll/win32/shlwapi/autocomp.cpp [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    const DWORD dwClsCtx = CLSCTX_INPROC_HANDLER | CLSCTX_INPROC_SERVER | CLSCTX_LOCAL_SERVER;
+    HRESULT hr = CoCreateInstance(CLSID_AutoComplete, NULL, dwClsCtx,
+                                  IID_IAutoComplete, (LPVOID *)&pAC2);
+    if (FAILED(hr))
+    {
+        ERR("CoCreateInstance(CLSID_AutoComplete) failed: 0x%lX\n", hr);
+        return hr;
+    }
+
+    CComPtr<IEnumString> pES;
```
Comment (learn-more) [inline dll/win32/shlwapi/autocomp.cpp:385]: `CComObject<IEnumString> ES`
or `CComObject<IEnumString>* pES`

That implements the QueryInterface and family

File: dll/win32/shlwapi/autocomp.cpp [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        : m_istr(0), m_hwndEdit(hwndEdit), m_dwSHACF(dwSHACF)
+    {
+        Reset();
+    }
+
+    CAutoCompleteEnumString(const CAutoCompleteEnumString& another)
+        : m_istr(another.m_istr)
+        , m_hwndEdit(another.m_hwndEdit)
+        , m_dwSHACF(another.m_dwSHACF)
+        , m_strs(another.m_strs)
```
Comment (learn-more) [inline dll/win32/shlwapi/autocomp.cpp:37]: This is used for `Clone`?
Does that also need to be bound to the same edit control?

File: dll/win32/shlwapi/autocomp.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp]: OK.

File: dll/win32/shlwapi/autocomp.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp]: Fixed to NULL.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_21_09_2020_19_28_18
How can I initialize `_pAtlModule`?

File: dll/win32/shlwapi/autocomp.cpp [hunk 8]
```diff
@@ -105,6 +112,9 @@ CAutoCompleteEnumString::Next(ULONG celt, LPOLESTR *rgelt, ULONG *pceltFetched)
 
 STDMETHODIMP CAutoCompleteEnumString::Skip(ULONG celt)
 {
+    if (m_bPending)
```
Comment (learn-more) [inline dll/win32/shlwapi/autocomp.cpp:123]: Should we wait for the result here?

Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp:123]: I think no.

**Comment by katahiromz** [comment]
Ready to go!

File: dll/win32/shlwapi/autocomp.cpp [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+    return FALSE;
+}
+
+static DWORD GetDriveFlags(LPCWSTR pszRoot)
+{
+    WCHAR szDevice[16];
+    HANDLE hDevice;
+    ULONG ret = 0;
+
+    lstrcpynW(szDevice, L"\\\\.\\", _countof(szDevice));
```
Comment (learn-more) [inline dll/win32/shlwapi/autocomp.cpp:169]: I would prefer String[Cch|Cb] functions in new code.

File: dll/win32/shlwapi/autocomp.cpp [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+                DoDir(pszQuery, bDirOnly); // Scan the directory
+                return FALSE; // Not exact match
+            }
+        }
+        return TRUE; // Exact match should hide the list
+    }
+    else if (pszQuery[0] && wcschr(pszQuery, L'\\') != NULL)
+    {
+        // Non-existent but can be a partial path
+        WCHAR szPath[MAX_PATH];
```
Comment (learn-more) [inline dll/win32/shlwapi/autocomp.cpp:237]: If you would use a CString here, that would remove the MAX_PATH limitation.

File: dll/win32/shlwapi/autocomp.cpp (hunk 9 above)
Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp]: OK.

File: dll/win32/shlwapi/autocomp.cpp (hunk 10 above)
Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp]: OK.

File: dll/win32/shlwapi/autocomp.cpp [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+    }
+
+    return FALSE; // Not exact match
+}
+
+void CAutoCompleteEnumString::DoTypedPaths(LPCWSTR pszQuery)
+{
+    static const LPCWSTR
+        pszTypedPaths = L"Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\TypedPaths";
+    WCHAR szName[32], szValue[MAX_PATH + 32];
```
Comment (katahiromz) [inline dll/win32/shlwapi/autocomp.cpp:258]: We should remove `MAX_PATH` of here?

Comment (learn-more) [reply dll/win32/shlwapi/autocomp.cpp:258]: If it is not making this very complex, yes please.

Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp]: OK. Done.

File: dll/win32/shlwapi/autocomp.cpp [hunk 12]
```diff
@@ -251,11 +251,37 @@ BOOL CAutoCompleteEnumString::DoFileSystem(LPCWSTR pszQuery)
     return FALSE; // Not exact match
 }
 
+static LONG
```
Comment (HBelusca) [inline dll/win32/shlwapi/autocomp.cpp:254]: This smells like something a CRegKey object could do.

Comment (katahiromz) [reply dll/win32/shlwapi/autocomp.cpp:254]: Used `CRegKey`. OK.

**Comment by katahiromz** [comment]
Closed because real implementation is quite different. Sorry for worrying a long time.

**Pull request closed without merging**